File: mediantwoarrays.java
//Time complexity:O(log m) m=smaller array
//Space complexity:O(1)

class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        
        int n1=nums1.length;
        int n2=nums2.length;
        if(n1>n2){
            return findMedianSortedArrays(nums2,nums1);
        }
        int low=0;
        int high=nums1.length;
        while(low<=high){
            int partX=low+(high-low)/2;
            int partY=(n1+n2+1)/2 - partX;
            double l1=Integer.MIN_VALUE;
            double l2=Integer.MIN_VALUE;
            double R1=Integer.MAX_VALUE;
            double R2=Integer.MAX_VALUE;
            if(partX!=0){
                l1=nums1[partX-1];
            }
            if(partX!=nums1.length){
                R1=nums1[partX];
            }
            if(partY!=0){
                l2=nums2[partY-1];
            }
            if(partY!=nums2.length){
                R2=nums2[partY];
            }
            
            if(l1<=R2 && l2<=R1){
                if((n1+n2)%2!=0){
                    System.out.println("odd" + l1 + " " + l2 + " " + R1 + " " + R2);
                    return Math.max(l1,l2);
                }
                else{
                    System.out.println("even" + l1 + " " + l2 + " " + R1 + " " + R2);
                    return (Math.max(l1,l2)+Math.min(R1,R2))/2;
                }
            }
            else if(l1>R2){
                high=partX-1;
            }
            else{
                low=partX+1;
            }
        }
    
    return -1;
    }
}

File: MedianTwoArrays.cpp
// Time : O(log(min(n1,n2)))
// Space : O(1)

class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        int n1 = nums1.size();
        int n2 = nums2.size();
        
        if (n1 == 0 && n2 == 0) {
            return 0.0;
        }
        
       
        if (n2 < n1) {
            return findMedianSortedArrays(nums2, nums1);
        }
        
        int low = 0;
        int high = n1;
        
        while (low <= high) {
            int PartX = low + (high - low) / 2;
            int PartY = (n1 + n2) / 2 - PartX;
            
            double L1 = PartX == 0 ? INT_MIN : nums1[PartX - 1];
            double R1 = PartX == n1 ? INT_MAX : nums1[PartX];
            double L2 = PartY == 0 ? INT_MIN: nums2[PartY - 1];
            double R2 = PartY == n2 ? INT_MAX : nums2[PartY];
            
            if (L1 <= R2 && L2 <= R1) {
                if ((n1 + n2) % 2 != 0) {
                    return min(R1, R2);
                } else {
                    return (max(L1, L2) + min(R1, R2)) / 2;
                }
            } else if (L2 > R1) {
                low = PartX + 1;
            } else {
                high = PartX - 1;
            }
        }
        
        return 8928903.39084;
        
        
        
    }
};